File: Analizador Semantico/Analizador/src/Postfija.java
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Stack;

public class Postfix {
    private List<Token> infija;
    private Stack<Token> stackk;
    private List<Token> postfija;
    private Map<String, TokenType> palabras_reservadas;

    public Postfix(List<Token> tokens) {
        this.infija = tokens;
        this.stackk = new Stack<>();
        this.postfija = new ArrayList<>();
        this.palabras_reservadas = new HashMap<>();

        this.palabras_reservadas.put("class", TokenType.CLASS);
        this.palabras_reservadas.put("also", TokenType.ALSO);
        this.palabras_reservadas.put("for", TokenType.FOR);
        this.palabras_reservadas.put("fun", TokenType.FUN);
        this.palabras_reservadas.put("if", TokenType.IF);
        this.palabras_reservadas.put("null", TokenType.NULL);
        this.palabras_reservadas.put("print", TokenType.PRINT);
        this.palabras_reservadas.put("return", TokenType.RETURN);
        this.palabras_reservadas.put("super", TokenType.SUPER);
        this.palabras_reservadas.put("this", TokenType.THIS);
        this.palabras_reservadas.put("var", TokenType.VAR);
        this.palabras_reservadas.put("while", TokenType.WHILE);
        this.palabras_reservadas.put("else", TokenType.ELSE);
    }

    public List<Token> convertir() {
        boolean estructuraDeControl = false;
        Stack<Token> stack_Estruc = new Stack<>();
        int auxfor = 0;
        for (int index = 0; index < this.infija.size(); index++) {
            Token t = this.infija.get(index);
            if (t.getType() == TokenType.EOF) {
                break;
            }
            if (this.palabras_reservadas.containsValue(t.getType())) {
                this.postfija.add(t);
                if (this.esEstructuraDeControl(t.getType())) {
                    estructuraDeControl = true;
                    stack_Estruc.push(t);
                }
            } else if (this.esOperando(t.getType())) {
                this.postfija.add(t);
            } else if (t.getType() == TokenType.PARENT_OPEN) {
                this.stackk.push(t);
            } else if (t.getType() == TokenType.PARENT_CLOSE) {
                while (!this.stackk.isEmpty() && this.stackk.peek().getType() != TokenType.PARENT_OPEN) {
                    Token temp = this.stackk.pop();
                    this.postfija.add(temp);
                }
                if (!this.stackk.isEmpty() && this.stackk.peek().getType() == TokenType.PARENT_OPEN) {
                    this.stackk.pop();
                }
                if (estructuraDeControl) {
                    this.postfija.add(new Token(TokenType.SEMICOLON, ";", ";", null));
                }
            } else if (this.esOperador(t.getType())) {
                while (!this.stackk.isEmpty() && this.precedenciaMayorIgual(this.stackk.peek().getType(), t.getType())) {
                    Token temp = this.stackk.pop();
                    this.postfija.add(temp);
                }
                this.stackk.push(t);
            } else if (t.getType() == TokenType.SEMICOLON) {
                while (!this.stackk.isEmpty() && this.stackk.peek().getType() != TokenType.BRACKET_OPEN) {
                    if (estructuraDeControl) {
                        if (stack_Estruc.peek().getType() == TokenType.FOR) {
                            Token temp = this.stackk.pop();
                            this.postfija.add(temp);
                            break;
                        }
                        continue;
                    }
                    Token temp = this.stackk.pop();
                    this.postfija.add(temp);
                }
                this.postfija.add(t);
            } else if (t.getType() == TokenType.BRACKET_OPEN) {
                this.stackk.push(t);
            } else if (t.getType() == TokenType.BRACKET_CLOSE && estructuraDeControl) {
                if (this.infija.get(index + 1).getType() == TokenType.ELSE) {
                    this.stackk.pop();
                } else {
                    this.stackk.pop();
                    this.postfija.add(new Token(TokenType.SEMICOLON, ";", ";", null));
                    stack_Estruc.pop();
                    if (stack_Estruc.isEmpty()) {
                        estructuraDeControl = false;
                    }
                }
            }
        }
        while (!this.stackk.isEmpty()) {
            Token temp = this.stackk.pop();
            this.postfija.add(temp);
        }
        while (!stack_Estruc.isEmpty()) {
            stack_Estruc.pop();
            this.postfija.add(new Token(TokenType.SEMICOLON, ";", ";", null));
        }
        return this.postfija;
    }

    private boolean esOperando(TokenType tipo) {
        switch (tipo) {
            case IDENTIFIER:
                return true;
            case NUMBER:
            case STRING:
            case TRUE:
            case FALSE:
                return true;
            default:
                return false;
        }
    }

    private boolean esEstructuraDeControl(TokenType tipo) {
        switch (tipo) {
            case IF:
            case ELSE:
                return true;
            case WHILE:
            case FOR:
                return true;
            default:
                return false;
        }
    }

    private boolean esOperador(TokenType tipo) {
        switch (tipo) {
            case ADD:
            case SUB:
            case MULT:
            case DIAG:
                return true;
            case EQUAL:
            case DIFERENT:
            case GREAT:
            case GREAT_EQUAL:
                return true;
            case AND:
            case OR:
            case ASIGNATION:
                return true;
            case LESS_THAN:
            case LESS_EQUAL:
                return true;
            default:
                return false;
        }
    }

    private boolean precedenciaMayorIgual(TokenType tipo1, TokenType tipo2) {
        return this.obtenerPrecedencia(tipo1) >= this.obtenerPrecedencia(tipo2);
    }

    private int obtenerPrecedencia(TokenType tipo) {
        switch (tipo) {
            case MULT:
            case DIAG:
                return 7;
            case ADD:
            case SUB:
                return 6;
            case GREAT_EQUAL:
            case GREAT:
            case LESS_THAN:
            case LESS_EQUAL:
                return 5;
            case DIFERENT:
            case EQUAL:
                return 4;
            case AND:
                return 3;
            case OR:
                return 2;
            case ASIGNATION:
                return 1;
            default:
                return 0;
        }
    }

    private int obAridad(TokenType tipo) {
        switch (tipo) {
            case MULT:
            case DIAG:
            case SUB:
            case ADD:
            case EQUAL:
            case GREAT:
            case GREAT_EQUAL:
            case ASIGNATION:
                return 2;
            case LESS_THAN:
            case LESS_EQUAL:
            case DIFERENT:
            case AND:
            case OR:
                return 2;
            default:
                return 0;
        }
    }
}

File: Analizador Semantico/Postfija.py
from Token import Token
from Tokentype import TokenType
from SymbolTable import TSimbolos

class Postfix:
    def __init__(self, tokens) -> None:
        #self.simbolos = simbolos
        self.infija = tokens
        self.stackk = []
        self.postfija = []
        self.palabras_reservadas = {  # Diccionario con las palabras reservadas y su tipo de token correspondiente
            # "palabra" - "tipo token"
            "class": TokenType.CLASS,
            "also": TokenType.ALSO,
            "for": TokenType.FOR,
            "fun": TokenType.FUN,
            "if": TokenType.IF,
            "null": TokenType.NULL,
            "print": TokenType.PRINT,
            "return": TokenType.RETURN,
            "super": TokenType.SUPER,
            "this": TokenType.THIS,
            #"true": TokenType.TRUE,
            "var": TokenType.VAR,
            "while": TokenType.WHILE,
            "else": TokenType.ELSE,
            #"true": TokenType.TRUE,
            #"false": TokenType.FALSE,
            #"and": TokenType.AND,
            #"or": TokenType.OR
        }

    def convertir(self):
        estructuraDeControl = False
        stack_Estruc:list[Token] = []
        auxfor = 0

        for index,t in enumerate(self.infija):
            if t.type == TokenType.EOF:
                break
            
            if t.type in self.palabras_reservadas.values():
                self.postfija.append(t)
                if self.esEstructuraDeControl(t.type):
                    estructuraDeControl = True
                    stack_Estruc.append(t)
            elif self.esOperando(t.type):
                self.postfija.append(t)
            elif t.type == TokenType.PARENT_OPEN:
                self.stackk.append(t)
            elif t.type == TokenType.PARENT_CLOSE:
                #print(f"{estructuraDeControl} y {self.stackk[-1].type}")
                while( ( len(self.stackk) != 0) and (self.stackk[-1].type != TokenType.PARENT_OPEN) ):
                    #if estructuraDeControl:
                        #if stack_Estruc[-1].type == TokenType.FOR:
                    temp = self.stackk.pop()
                    self.postfija.append(temp)
                if self.stackk[-1].type == TokenType.PARENT_OPEN:
                    self.stackk.pop()
                if estructuraDeControl:
                    #print("semiak")
                    self.postfija.append(Token(TokenType.SEMICOLON,";",";",None))
            elif self.esOperador(t.type):
                #print(f"si -> {t.type}")
                while len(self.stackk)!=0 and self.precedenciaMayorIgual(self.stackk[-1].type, t.type):
                    temp = self.stackk.pop()
                    self.postfija.append(temp)
                self.stackk.append(t)
            elif t.type == TokenType.SEMICOLON:
                while len(self.stackk) !=0 and self.stackk[-1].type != TokenType.BRACKET_OPEN:
                    if estructuraDeControl:
                        if stack_Estruc[-1].type == TokenType.FOR: #and auxfor <2:
                            temp = self.stackk.pop()
                            self.postfija.append(temp)
                            break
                        pass
                    temp = self.stackk.pop()
                    self.postfija.append(temp)
                self.postfija.append(t) 
            elif t.type == TokenType.BRACKET_OPEN:
                self.stackk.append(t)
            elif t.type == TokenType.BRACKET_CLOSE and estructuraDeControl:
                #print(f"caso else -> {self.infija[index+1].type}")
                if self.infija[index+1].type == TokenType.ELSE:
                    self.stackk.pop()
                else:
                    self.stackk.pop()
                    #print("aki")
                    self.postfija.append(Token(TokenType.SEMICOLON,";",";",None))
                    #self.postfija.append("semi")
                    stack_Estruc.pop()
                    if len(stack_Estruc) == 0:
                        estructuraDeControl = False

        while(len(self.stackk) != 0):
            temp = self.stackk.pop()
            self.postfija.append(temp)

        while( len(stack_Estruc) != 0):
            #print("sem")
            stack_Estruc.pop()
            self.postfija.append(Token(TokenType.SEMICOLON,";",";",None))


        return self.postfija

    def esOperando(self,tipo:TokenType):
        match tipo:
            case TokenType.IDENTIFIER:
                return True
            case TokenType.NUMBER | TokenType.STRING | TokenType.TRUE | TokenType.FALSE:
                return True
            case other:
                return False

    def esEstructuraDeControl(self,tipo):
        match tipo:
            case TokenType.IF | TokenType.ELSE:
                return True
            case TokenType.WHILE | TokenType.FOR:
                return True
            case _:
                return False

    def esOperador(self,tipo:TokenType):
        #print("esop")
        match tipo:
            case TokenType.ADD | TokenType.SUB | TokenType.MULT | TokenType.DIAG:
                return True
            case TokenType.EQUAL | TokenType.DIFERENT | TokenType.GREAT | TokenType.GREAT_EQUAL:
                return True
            case TokenType.AND | TokenType.OR | TokenType.ASIGNATION:
                return True
            case TokenType.LESS_THAN | TokenType.LESS_EQUAL:
                return True
            case other:
                return False


    def precedenciaMayorIgual(self, tipo1, tipo2):
        #print(f"{tipo1}->{self.obtenerPrecedencia(tipo1)} y  {tipo2}->{self.obtenerPrecedencia(tipo2)}")
        return  self.obtenerPrecedencia(tipo1) >= self.obtenerPrecedencia(tipo2)

    def obtenerPrecedencia(self,tipo):
        match tipo:
            case TokenType.MULT | TokenType.DIAG:
                return 7 
            case TokenType.ADD | TokenType.SUB:
                return 6
            case TokenType.GREAT_EQUAL | TokenType.GREAT | TokenType.LESS_THAN | TokenType.LESS_EQUAL:
                return 5
            case TokenType.DIFERENT | TokenType.EQUAL:
                return 4
            case TokenType.AND:
                return 3
            case TokenType.OR:
                return 2
            case TokenType.ASIGNATION:
                return 1
            case _:
                return 0


    def obAridad(self,tipo: TokenType):
        match tipo:
            case TokenType.MULT| TokenType.DIAG| TokenType.SUB| TokenType.ADD| TokenType.EQUAL| TokenType.GREAT| TokenType.GREAT_EQUAL | TokenType.ASIGNATION:
                return 2
            case TokenType.LESS_THAN | TokenType.LESS_EQUAL | TokenType.DIFERENT | TokenType.AND | TokenType.OR :
                return 2
            case other:
                return 0